types/react-relay/legacy.d.ts

```diff
@@ -0,0 +1,143 @@
+import * as React from 'react';
+
+import {
+    IEnvironment,
+    Variables,
+    CacheConfig,
+    GraphQLTaggedNode,
+    RelayContext,
+    PageInfo,
+    OperationType,
+    _FragmentRefs,
+    _RefType,
+    FetchPolicy,
+} from 'relay-runtime';
+
+export {
+    FragmentRef,
+    RelayPaginationProp,
+    RelayProp,
+    RelayRefetchProp,
+} from './ReactRelayTypes';
+
+import { RelayProp, MappedFragmentProps, RelayRefetchProp, RelayPaginationProp } from './ReactRelayTypes';
+
+export {
+    DataID,
+    DeclarativeMutationConfig,
+    Disposable,
+    // RelayRuntime has two environment exports: one interface, one concrete.
+    IEnvironment as Environment,
+    GraphQLTaggedNode,
+    MutationType,
+    NormalizationSelector,
+    OperationDescriptor,
+    RangeOperation,
+    ReaderSelector,
+    RelayContext,
+    Snapshot,
+    Variables,
+    FetchPolicy,
+} from 'relay-runtime';
+
+/**
+ * Legacy react-relay exports.
+ * Should prefer using interface defined in ./hooks.js
+ */
+export { ConnectionHandler } from 'relay-runtime';
+interface QueryRendererProps<TOperation extends OperationType> {
+    environment: IEnvironment;
+    query: GraphQLTaggedNode | null | undefined;
+    render: (renderProps: {
+        error: Error | null;
+        props: TOperation['response'] | null;
+        retry: (() => void) | null;
+    }) => React.ReactNode;
+    variables: TOperation['variables'];
+}
+declare class ReactRelayQueryRenderer<TOperation extends OperationType> extends React.Component<
+    {
+        cacheConfig?: CacheConfig | null;
+        fetchPolicy?: FetchPolicy;
+    } & QueryRendererProps<TOperation>
+    > { }
+export { ReactRelayQueryRenderer as QueryRenderer };
+
+declare class ReactRelayLocalQueryRenderer<TOperation extends OperationType> extends React.Component<
+    QueryRendererProps<TOperation>
+    > { }
+export { ReactRelayLocalQueryRenderer as LocalQueryRenderer };
+
+export { MutationTypes } from 'relay-runtime';
+export { RangeOperations } from 'relay-runtime';
+
+export const ReactRelayContext: React.Context<RelayContext | null>;
+
+export { applyOptimisticMutation } from 'relay-runtime';
+export { commitLocalUpdate } from 'relay-runtime';
+export { commitMutation } from 'relay-runtime';
+
+export type ContainerProps<Props> = MappedFragmentProps<Pick<Props, Exclude<keyof Props, 'relay'>>>;
+export type RelayProps<Props> = ContainerProps<Props>; // TODO: validate this
```

Comment:
@rbalicki2 can you just help me and check if this is correct? Struggling to understand the flow equivalent to this one

Reply:
# Flow

As far as I can tell, Props2 = $RelayProps<Props1, RelayProp> is generating a new type which is like Props1, except:
* Excluding `{relay: RelayProp}`, and
* Mapping all remaining props which are generated GraphQLTaggedLiterals to the value of their $refType field (preserving things like nullability and whether it's an array), and
* leaving all remaining props alone

I think!

So, it is used in e.g. `createContainerWithProps` to create a new component type (Component2) to which you pass Props2, i.e.:

* don't pass `relay`
* pass `$key` types (e.g. FooFragment_user$key, I think)

And which reverses this, i.e. adds `relay` and looks up fragments by their ref (via RelayModernFragmentSpecResolver), and calls `Component1` with `Props1`.

# Typescript

That looks like what is happening in Typescript as well, but it's hard to be entirely sure.

Maybe it would be good to check whether all props are correctly mapped? e.g. making sure that all `FragmentOrRegularProp` matches the behavior of $RelayProps.

In particular, I would make sure that if the array is nullable, FragmentOrRegularProp applies. That seems to not be covered by FragmentOrRegularProp. But it's not clear!

See this: https://www.typescriptlang.org/play?#code/FAFwngDgpgBAgjAvDA3gQwFwxAJwK5QC+A3MKJLAEJLwwA+MAbgPYCWAJqedDAMIA8AFQB8NQTCgAPEFAB27AM60A-KlYLMMWXgC2AIyg5CMLCnV6s2-YZJlwPACI0BlYVwD07mE4UALZngANuwwAMaBUGg4gWASrCC+hjAGahqWugZGMMw4qRZaGTYANMl4IDDxAORKeswJysp2FDAAojQAFAgMLBwAlBLScooq2PiwWABmaIEKUKRAA
Type E seems to indicate that A | void doesn't extend A, but Typescript gets confused with type D.

Nullable arrays don't show up in Relay, though, so I think you're probably good.

# Conclusion

Did this work previously, with no complaints? If this is broken, it seems like it would have tripped someone else up.

I would land this, and maybe add some tests as a follow up?